lib/io.dx

```diff
@@ -0,0 +1,77 @@
+
+'File system operations
+
+FilePath : Type = String
+data CString = MkCString CharPtr
+
+data StreamMode =
+  ReadMode
+  WriteMode
+
+data Stream mode:StreamMode = MkStream CharPtr
+
+-- TODO: check the string contains no nulls
+def withCString (s:String) (action: CString -> {State World} a) : {State World} a =
+  (AsList n s') = s <> (AsList _ "\NUL")
+  withTabPtr s' \(MkPtr ptr). action $ MkCString ptr
+
+def fopen (path:String) (mode:StreamMode) : {State World} (Stream mode) =
+  modeStr = AsList _ case mode of
+    ReadMode  -> "r"
+    WriteMode -> "w"
+  withCString path \(MkCString pathPtr).
+    withCString modeStr \(MkCString modePtr).
+      MkStream $ %ffi fopen CharPtr pathPtr modePtr
+
+def fclose (mode:StreamMode) ?-> (stream:Stream mode) : {State World} Unit =
+  (MkStream stream') = stream
+  %ffi fclose Int64 stream'
+  ()
+
+def fwrite (stream:Stream WriteMode) (s:String) : {State World} Unit =
+  (MkStream stream') = stream
+  (AsList n s') = s
+  withTabPtr s' \(MkPtr ptr).
+    %ffi fwrite Int64 ptr (IToI64 1) (IToI64 n) stream'
+  ()
+
+def fread (stream:Stream ReadMode) : {State World} String =
+  (MkStream stream') = stream
+  -- TODO: allow reading longer files!
+  n = 4096
+  withAlloc n \ptr:(Ptr Char).
+    (MkPtr rawPtr) = ptr
+    numRead = I64ToI $ %ffi fread Int64 rawPtr (IToI64 1) (IToI64 n) stream'
+    stringFromCharPtr numRead ptr
+
+def deleteFile (f:FilePath) : {State World} Unit =
+  withCString f \(MkCString ptr).
+    %ffi remove Int64 ptr
+  ()
+
+def withFile (f:FilePath) (mode:StreamMode)
+      (action: Stream mode -> {State World} a)
+      : {State World} a =
+  stream = fopen f mode
+  result = action stream
+  fclose stream
+  result
+
+def writeFile (f:FilePath) (s:String) : {State World} Unit =
+  withFile f WriteMode \stream. fwrite stream s
+
+def readFile (f:FilePath) : {State World} String =
+  withFile f ReadMode \stream. fread stream
+
+def writeTemp (s:String) : {State World} FilePath =
+  -- TODO: Make this properly atomic. It can fail if another process creates a
```

Comment:
Maybe use `mkstemp` instead?

Reply:
Good idea. When I first looked at it I thought that it gave you an FD but no file name (which we need for shelling out) but actually it gives you a file name too by modifying the buffer in-place.